agent/engine/task_manager.go

```diff
@@ -415,8 +415,15 @@ func (task *managedTask) progressContainers() {
 	task.UpdateStatus()
 }
 
-func (task *managedTask) cleanupTask() {
-	cleanupTime := ttime.After(task.KnownStatusTime.Add(taskStoppedDuration).Sub(ttime.Now()))
+func (task *managedTask) cleanupTask(taskStoppedDuration time.Duration) {
+	cleanupTimeDuration := task.KnownStatusTime.Add(taskStoppedDuration).Sub(ttime.Now())
+	// There is a potential deadlock here if cleanupTime is negative. Ignore the computed
+	// value in this case in favor of the default config value.
```

Comment:
A negative here can actually happen in normal scenarios as well, counter to what you say above @aaithal.

If you stop the agent while a task is waiting for cleanup for a duration longer than `taskStoppedDuration`, I think you reliably end up hitting the default value here. e.g. if your stoppedDuration is 1m and the task stops at 00:00, you stop the agent right after, and start it at 00:10, you get 00:00 + 00:01 - 00:10 = -00:09 (the task should have been removed 9 minutes ago, 1 minute after it stopped).

In that case, the right thing is to immediately clean it up, not wait another 3 hours and then clean it up.

I might be misunderstanding, so do correct me if I'm wrong.

I think we clearly want to root cause the deadlock observed here, (hopefully it's #313 and fixing that'll let us remove this) and get this check out of here. if we can.